petab_select/candidate_space.py

```diff
@@ -529,6 +602,628 @@ def iterate():
+        )
+
+        for swap_set in self.swap_parameter_sets:
+            if swapped_parameters_ids.issubset(set(swap_set)):
+                return True
+        return False
+
+    def check_critical(self, model: Model) -> bool:
+        """Check if the model contains all necessary critical parameters"""
+
+        estimated_parameters_ids = set(model.get_estimated_parameter_ids_all())
+        for critical_set in self.critical_parameter_sets:
+            if not estimated_parameters_ids.intersection(set(critical_set)):
+                return False
+        return True
+
+    def switch_method(self) -> None:
+        """Switch to the next method with respect to the history
+        of methods used and the switching scheme in self.method_scheme"""
+
+        previous_method = self.method
+        next_method = previous_method
+        logging.info("SWITCHING", self.method_history)
+        # breakpoint()
+        # If last method was LATERAL and we have made a succesfull swap
+        # (a better model was found) then go back to FORWARD. Else do the
+        # usual method swapping scheme.
+        if self.swap_done_successfully:
+            next_method = self.method_scheme[(Method.LATERAL,)]
+        else:
+            # iterate through the method_scheme dictionary to see which method to switch to
+            for previous_methods in self.method_scheme:
+                if previous_methods is not None and previous_methods == tuple(
+                    self.method_history[-len(previous_methods) :]
+                ):
+                    next_method = self.method_scheme[previous_methods]
+                    # if found a switch just break (choosing first good switch)
+                    break
+
+        # raise error if the method didn't change
+        if next_method == previous_method:
+            raise ValueError(
+                "Method didn't switch when it had to. "
+                "The `method_scheme` provided is not sufficient. "
+                "Please provide a correct method_scheme scheme. "
+                f"Method history: `{self.method_history}`. "
+            )
+
+        # if the next method is None (in default case if SWAP
+        # method didn't find any better models) then terminate
+        if not next_method:
+            if self.number_of_reattempts:
+                self.jump_to_most_distant()
+                return
+            else:
+                raise StopIteration(
+                    f"The next method provided is None. The search is terminating."
+                )
+
+        # If we try to switch to SWAP method but it's not available (no crit or swap parameter groups)
+        if (
+            next_method == Method.LATERAL
+            and not [
+                critical_set
+                for critical_set in self.critical_parameter_sets
+                if len(critical_set) > 1
+            ]
+            and not self.swap_parameter_sets
+        ):
```

Comment:
Critical sets are checked here as a possibility for swaps, but not in `check_swap`. Should they be checked there too?

Reply:
So in original famos critical sets were also swap sets. So in a swap could've happened in critical sets as well. 
I put this when I was first making the switch same to original famos. But then later when I wrote swap I thought it would make more sense/be clearer to have only swap sets be sets in which one could swap. 

So I would say that we can remove the critical sets here. But I can also add critical sets to the swapping.